{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Experiment 2 - Simulated mutated references, simulate raw data, align raw data to mutated references\n",
    "\n",
    "This notebook is modified from the original experiment-2 version to deal with proteins!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import numpy as np\n",
    "import random\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import time\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read the original NCBI reference sequence\n",
    "\n",
    "#filename = \"../refs/rhinovirus_YP_001552411.1.fasta\" \n",
    "#filename = \"../refs/streptococcus_VDG79570_102aa.fasta\"\n",
    "filename = \"../refs/streptococcus_VDG79582_PBP_731aa.fasta\"\n",
    "#filename = \"../refs/streptococcus_VDG79577_shortened_50aa.fasta\"\n",
    "\n",
    "original_record = SeqIO.read(filename, \"fasta\") # NOTE: this only works with single-sequence .fasta files\n",
    "\n",
    "#id_thresholds = [0.99, 0.97, 0.95, 0.92, 0.9, 0.85, 0.8]  # specified list of thresholds\n",
    "id_thresholds = [0.9, 0.85, 0.8, 0.75, 0.7, 0.65, 0.6]\n",
    "\n",
    "def mutate_sequence(sequence, pid):\n",
    "    '''\n",
    "    Given a sequence and a % ID threshold, mutate the sequence at random positions to produce\n",
    "    a new sequence with the specified % ID. Return the new sequence string.\n",
    "    '''\n",
    "    \n",
    "    swap_values = ['A', 'R', 'N', 'D', 'C', 'Q', 'E', 'G', 'H', 'I', 'L', 'K', \n",
    "                             'M', 'F', 'P', 'S', 'T', 'W', 'Y', 'V'] # amino acid characters\n",
    "    \n",
    "    len_of_seq = len(sequence)\n",
    "    num_bp_to_change = (np.round(len_of_seq) * (1-pid))\n",
    "    all_bp_locs = [b for b in range(len_of_seq)]    \n",
    "    bp_locs_to_change = random.sample(all_bp_locs, int(num_bp_to_change))\n",
    "    \n",
    "    # loop through indices of bases to change and swap out values\n",
    "    new_seq = list(sequence)\n",
    "    for bp in bp_locs_to_change:\n",
    "        new_seq[bp] = random.sample([i for i in swap_values if i != new_seq[bp] ], 1)[0]\n",
    "        \n",
    "    return(\"\".join(new_seq))\n",
    "\n",
    "# currate a dictionary of mutated sequences at specified %ID thresholds\n",
    "sequence_dict = {original_record.id : original_record.seq}\n",
    "for i in id_thresholds:\n",
    "    for j in range(1):\n",
    "        this_seq = mutate_sequence(original_record.seq, i)\n",
    "        #this_seq_rt = protein_to_dna(this_seq)\n",
    "        sequence_dict[str(i) + '-' + str(j+1)] = this_seq#_rt\n",
    "        \n",
    "# write the mutated sequences to a fasta file\n",
    "for s in sequence_dict:\n",
    "    with open( s + \".ref.fasta\", 'w') as f: \n",
    "        f.write('>' + s + '\\n' + str(sequence_dict[s]) + '\\n')\n",
    "    f.close()\n",
    "    \n",
    "#NOTE: verified that this is working correctly by running BLASTx on the resulting *.ref.fasta sequences\n",
    "#      the BLASTx results for the NT sequence showed that the NT sequence listed as 99% has 99% similarity, \n",
    "#      the NT seq for 97% has 97% protein similarity...etc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "def protein_to_dna(protein_sequence):\n",
    "    '''\n",
    "    Given a protein sequence, reverse translate the sequence to produce a nucleotide sequence\n",
    "    '''\n",
    "    # Dictionary mapping amino acids to DNA codons\n",
    "    codon_table = {\n",
    "        'A': ['GCT', 'GCC', 'GCA', 'GCG'],\n",
    "        'C': ['TGT', 'TGC'],\n",
    "        'D': ['GAT', 'GAC'],\n",
    "        'E': ['GAA', 'GAG'],\n",
    "        'F': ['TTT', 'TTC'],\n",
    "        'G': ['GGT', 'GGC', 'GGA', 'GGG'],\n",
    "        'H': ['CAT', 'CAC'],\n",
    "        'I': ['ATT', 'ATC', 'ATA'],\n",
    "        'K': ['AAA', 'AAG'],\n",
    "        'L': ['TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG'],\n",
    "        'M': ['ATG'],\n",
    "        'N': ['AAT', 'AAC'],\n",
    "        'P': ['CCT', 'CCC', 'CCA', 'CCG'],\n",
    "        'Q': ['CAA', 'CAG'],\n",
    "        'R': ['CGT', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG'],\n",
    "        'S': ['TCT', 'TCC', 'TCA', 'TCG', 'AGT', 'AGC'],\n",
    "        'T': ['ACT', 'ACC', 'ACA', 'ACG'],\n",
    "        'V': ['GTT', 'GTC', 'GTA', 'GTG'],\n",
    "        'W': ['TGG'],\n",
    "        'Y': ['TAT', 'TAC'],\n",
    "        '*': ['TAA', 'TAG', 'TGA']  # Stop codons\n",
    "    }\n",
    "    \n",
    "    dna_sequence = ''\n",
    "    \n",
    "    # Convert each amino acid to the corresponding DNA codon\n",
    "    for amino_acid in protein_sequence:\n",
    "        codons = codon_table.get(amino_acid, [])\n",
    "        if len(codons) == 0:\n",
    "            raise ValueError(f\"Invalid amino acid: {amino_acid}\")\n",
    "        \n",
    "        # Choose a random codon for the amino acid\n",
    "        codon = codons[0]  # Change this if you want a random codon\n",
    "        \n",
    "        dna_sequence += codon\n",
    "    \n",
    "    return dna_sequence\n",
    "\n",
    "\n",
    "# write a nucleotide sequence version of the original input to file for use in data simulation step\n",
    "nt_filename = '.'.join(filename.split('/')[-1].split('.')[0:-1])  + '.ntseq.fasta'\n",
    "with open(nt_filename, 'w') as f:\n",
    "    f.write('>' + original_record.id + '_ntseq' + '\\n')\n",
    "    f.write(protein_to_dna(original_record.seq))\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO:iss.app:Starting iss generate\n",
      "INFO:iss.app:Using kde ErrorModel\n",
      "INFO:iss.util:Stitching input files together\n",
      "INFO:iss.app:Using lognormal abundance distribution\n",
      "INFO:iss.app:Using 2 cpus for read generation\n",
      "INFO:iss.app:Generating 100000 reads\n",
      "INFO:iss.app:Generating reads for record: VDG79582.1_ntseq\n",
      "INFO:iss.util:Stitching input files together\n",
      "INFO:iss.util:Stitching input files together\n",
      "INFO:iss.util:Cleaning up\n",
      "INFO:iss.app:Read generation complete\n"
     ]
    }
   ],
   "source": [
    "# simulate sequences from the input file\n",
    "! iss generate --genomes {nt_filename} --model miseq --output miseq_reads --n_reads 100000\n",
    "! cat miseq_reads_R1.fastq miseq_reads_R2.fastq > miseq_reads_R1R2.fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "diamond v2.1.6.160 (C) Max Planck Society for the Advancement of Science\n",
      "Documentation, support and updates available at http://www.diamondsearch.org\n",
      "Please cite: http://dx.doi.org/10.1038/s41592-021-01101-x Nature Methods (2021)\n",
      "\n",
      "#CPU threads: 16\n",
      "Scoring parameters: (Matrix=BLOSUM62 Lambda=0.267 K=0.041 Penalties=11/1)\n",
      "Database input file: VDG79582.1.ref.fasta\n",
      "Opening the database file...  [0s]\n",
      "Loading sequences...  [0s]\n",
      "Masking sequences...  [0s]\n",
      "Writing sequences...  [0s]\n",
      "Hashing sequences...  [0s]\n",
      "Loading sequences...  [0s]\n",
      "Writing trailer...  [0s]\n",
      "Closing the input file...  [0s]\n",
      "Closing the database file...  [0s]\n",
      "\n",
      "Database sequences  1\n",
      "  Database letters  731\n",
      "     Database hash  32e9422a9d9728be30ec7cf2ad64c3de\n",
      "        Total time  0.000000s\n",
      "diamond v2.1.6.160 (C) Max Planck Society for the Advancement of Science\n",
      "Documentation, support and updates available at http://www.diamondsearch.org\n",
      "Please cite: http://dx.doi.org/10.1038/s41592-021-01101-x Nature Methods (2021)\n",
      "\n",
      "#CPU threads: 16\n",
      "Scoring parameters: (Matrix=BLOSUM62 Lambda=0.267 K=0.041 Penalties=11/1)\n",
      "Temporary directory: \n",
      "#Target sequences to report alignments for: 25\n",
      "Opening the database...  [0s]\n",
      "Database: reference (type: Diamond database, sequences: 1, letters: 731)\n",
      "Block size = 2000000000\n",
      "Opening the input file...  [0.023s]\n",
      "Opening the output file...  [0s]\n",
      "Loading query sequences...  [0.437s]\n",
      "Masking queries...  [70.966s]\n",
      "Algorithm: Double-indexed\n",
      "Building query histograms...  [0.282s]\n",
      "Seeking in database...  [0s]\n",
      "Loading reference sequences...  [0s]\n",
      "Masking reference...  [0.002s]\n",
      "Initializing temporary storage...  [0.005s]\n",
      "Building reference histograms...  [0s]\n",
      "Allocating buffers...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 1/8, index chunk 1/4.\n",
      "Building reference seed array...  [0.001s]\n",
      "Building query seed array...  [0.078s]\n",
      "Computing hash join...  [0.008s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.105s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 1/8, index chunk 2/4.\n",
      "Building reference seed array...  [0.001s]\n",
      "Building query seed array...  [0.084s]\n",
      "Computing hash join...  [0.017s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.088s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 1/8, index chunk 3/4.\n",
      "Building reference seed array...  [0s]\n",
      "Building query seed array...  [0.086s]\n",
      "Computing hash join...  [0.018s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.098s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 1/8, index chunk 4/4.\n",
      "Building reference seed array...  [0s]\n",
      "Building query seed array...  [0.082s]\n",
      "Computing hash join...  [0.011s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.083s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 2/8, index chunk 1/4.\n",
      "Building reference seed array...  [0.001s]\n",
      "Building query seed array...  [0.081s]\n",
      "Computing hash join...  [0.012s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.09s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 2/8, index chunk 2/4.\n",
      "Building reference seed array...  [0.001s]\n",
      "Building query seed array...  [0.083s]\n",
      "Computing hash join...  [0.015s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.076s]\n",
      "Deallocating memory...  [0.001s]\n",
      "Processing query block 1, reference block 1/1, shape 2/8, index chunk 3/4.\n",
      "Building reference seed array...  [0s]\n",
      "Building query seed array...  [0.083s]\n",
      "Computing hash join...  [0.022s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.089s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 2/8, index chunk 4/4.\n",
      "Building reference seed array...  [0s]\n",
      "Building query seed array...  [0.079s]\n",
      "Computing hash join...  [0.01s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.086s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 3/8, index chunk 1/4.\n",
      "Building reference seed array...  [0.001s]\n",
      "Building query seed array...  [0.083s]\n",
      "Computing hash join...  [0.02s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.094s]\n",
      "Deallocating memory...  [0.001s]\n",
      "Processing query block 1, reference block 1/1, shape 3/8, index chunk 2/4.\n",
      "Building reference seed array...  [0.001s]\n",
      "Building query seed array...  [0.083s]\n",
      "Computing hash join...  [0.009s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.086s]\n",
      "Deallocating memory...  [0.001s]\n",
      "Processing query block 1, reference block 1/1, shape 3/8, index chunk 3/4.\n",
      "Building reference seed array...  [0.001s]\n",
      "Building query seed array...  [0.087s]\n",
      "Computing hash join...  [0.01s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.078s]\n",
      "Deallocating memory...  [0.001s]\n",
      "Processing query block 1, reference block 1/1, shape 3/8, index chunk 4/4.\n",
      "Building reference seed array...  [0s]\n",
      "Building query seed array...  [0.083s]\n",
      "Computing hash join...  [0.01s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.117s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 4/8, index chunk 1/4.\n",
      "Building reference seed array...  [0.001s]\n",
      "Building query seed array...  [0.083s]\n",
      "Computing hash join...  [0.01s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.101s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 4/8, index chunk 2/4.\n",
      "Building reference seed array...  [0.001s]\n",
      "Building query seed array...  [0.084s]\n",
      "Computing hash join...  [0.009s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.086s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 4/8, index chunk 3/4.\n",
      "Building reference seed array...  [0s]\n",
      "Building query seed array...  [0.086s]\n",
      "Computing hash join...  [0.011s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.085s]\n",
      "Deallocating memory...  [0.001s]\n",
      "Processing query block 1, reference block 1/1, shape 4/8, index chunk 4/4.\n",
      "Building reference seed array...  [0.001s]\n",
      "Building query seed array...  [0.082s]\n",
      "Computing hash join...  [0.014s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.095s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 5/8, index chunk 1/4.\n",
      "Building reference seed array...  [0s]\n",
      "Building query seed array...  [0.088s]\n",
      "Computing hash join...  [0.021s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.099s]\n",
      "Deallocating memory...  [0.001s]\n",
      "Processing query block 1, reference block 1/1, shape 5/8, index chunk 2/4.\n",
      "Building reference seed array...  [0.001s]\n",
      "Building query seed array...  [0.083s]\n",
      "Computing hash join...  [0.012s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.09s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 5/8, index chunk 3/4.\n",
      "Building reference seed array...  [0s]\n",
      "Building query seed array...  [0.093s]\n",
      "Computing hash join...  [0.029s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.093s]\n",
      "Deallocating memory...  [0.002s]\n",
      "Processing query block 1, reference block 1/1, shape 5/8, index chunk 4/4.\n",
      "Building reference seed array...  [0.001s]\n",
      "Building query seed array...  [0.083s]\n",
      "Computing hash join...  [0.009s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.087s]\n",
      "Deallocating memory...  [0.001s]\n",
      "Processing query block 1, reference block 1/1, shape 6/8, index chunk 1/4.\n",
      "Building reference seed array...  [0.001s]\n",
      "Building query seed array...  [0.088s]\n",
      "Computing hash join...  [0.018s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.083s]\n",
      "Deallocating memory...  [0.001s]\n",
      "Processing query block 1, reference block 1/1, shape 6/8, index chunk 2/4.\n",
      "Building reference seed array...  [0s]\n",
      "Building query seed array...  [0.133s]\n",
      "Computing hash join...  [0.017s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.142s]\n",
      "Deallocating memory...  [0.001s]\n",
      "Processing query block 1, reference block 1/1, shape 6/8, index chunk 3/4.\n",
      "Building reference seed array...  [0.002s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Building query seed array...  [0.097s]\n",
      "Computing hash join...  [0.018s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.091s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 6/8, index chunk 4/4.\n",
      "Building reference seed array...  [0.001s]\n",
      "Building query seed array...  [0.092s]\n",
      "Computing hash join...  [0.017s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.139s]\n",
      "Deallocating memory...  [0.002s]\n",
      "Processing query block 1, reference block 1/1, shape 7/8, index chunk 1/4.\n",
      "Building reference seed array...  [0s]\n",
      "Building query seed array...  [0.083s]\n",
      "Computing hash join...  [0.01s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.094s]\n",
      "Deallocating memory...  [0.001s]\n",
      "Processing query block 1, reference block 1/1, shape 7/8, index chunk 2/4.\n",
      "Building reference seed array...  [0.001s]\n",
      "Building query seed array...  [0.085s]\n",
      "Computing hash join...  [0.019s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.1s]\n",
      "Deallocating memory...  [0.001s]\n",
      "Processing query block 1, reference block 1/1, shape 7/8, index chunk 3/4.\n",
      "Building reference seed array...  [0.001s]\n",
      "Building query seed array...  [0.095s]\n",
      "Computing hash join...  [0.025s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.116s]\n",
      "Deallocating memory...  [0.001s]\n",
      "Processing query block 1, reference block 1/1, shape 7/8, index chunk 4/4.\n",
      "Building reference seed array...  [0.002s]\n",
      "Building query seed array...  [0.083s]\n",
      "Computing hash join...  [0.022s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.082s]\n",
      "Deallocating memory...  [0.001s]\n",
      "Processing query block 1, reference block 1/1, shape 8/8, index chunk 1/4.\n",
      "Building reference seed array...  [0.001s]\n",
      "Building query seed array...  [0.086s]\n",
      "Computing hash join...  [0.015s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.098s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 8/8, index chunk 2/4.\n",
      "Building reference seed array...  [0s]\n",
      "Building query seed array...  [0.085s]\n",
      "Computing hash join...  [0.022s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.097s]\n",
      "Deallocating memory...  [0.001s]\n",
      "Processing query block 1, reference block 1/1, shape 8/8, index chunk 3/4.\n",
      "Building reference seed array...  [0.002s]\n",
      "Building query seed array...  [0.089s]\n",
      "Computing hash join...  [0.035s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.089s]\n",
      "Deallocating memory...  [0.002s]\n",
      "Processing query block 1, reference block 1/1, shape 8/8, index chunk 4/4.\n",
      "Building reference seed array...  [0.001s]\n",
      "Building query seed array...  [0.087s]\n",
      "Computing hash join...  [0.024s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.115s]\n",
      "Deallocating memory...  [0s]\n",
      "Deallocating buffers...  [0.001s]\n",
      "Clearing query masking...  [0.028s]\n",
      "Computing alignments... Loading trace points...  [0.027s]\n",
      "Sorting trace points...  [0.011s]\n",
      "Computing alignments...  [1.018s]\n",
      "Deallocating buffers...  [0s]\n",
      "Loading trace points...  [0s]\n",
      " [1.07s]\n",
      "Deallocating reference...  [0s]\n",
      "Loading reference sequences...  [0s]\n",
      "Deallocating buffers...  [0s]\n",
      "Deallocating queries...  [0.006s]\n",
      "Loading query sequences...  [0s]\n",
      "Closing the input file...  [0s]\n",
      "Closing the output file...  [0.007s]\n",
      "Closing the database...  [0s]\n",
      "Cleaning up...  [0s]\n",
      "Total time = 79.37s\n",
      "Reported 99963 pairwise alignments, 99963 HSPs.\n",
      "99963 queries aligned.\n",
      "diamond v2.1.6.160 (C) Max Planck Society for the Advancement of Science\n",
      "Documentation, support and updates available at http://www.diamondsearch.org\n",
      "Please cite: http://dx.doi.org/10.1038/s41592-021-01101-x Nature Methods (2021)\n",
      "\n",
      "#CPU threads: 16\n",
      "Scoring parameters: (Matrix=BLOSUM62 Lambda=0.267 K=0.041 Penalties=11/1)\n",
      "Database input file: 0.9-1.ref.fasta\n",
      "Opening the database file...  [0s]\n",
      "Loading sequences...  [0s]\n",
      "Masking sequences...  [0s]\n",
      "Writing sequences...  [0s]\n",
      "Hashing sequences...  [0s]\n",
      "Loading sequences...  [0s]\n",
      "Writing trailer...  [0s]\n",
      "Closing the input file...  [0s]\n",
      "Closing the database file...  [0s]\n",
      "\n",
      "Database sequences  1\n",
      "  Database letters  731\n",
      "     Database hash  754894dc132d89dd038f8f764e68099b\n",
      "        Total time  0.000000s\n",
      "diamond v2.1.6.160 (C) Max Planck Society for the Advancement of Science\n",
      "Documentation, support and updates available at http://www.diamondsearch.org\n",
      "Please cite: http://dx.doi.org/10.1038/s41592-021-01101-x Nature Methods (2021)\n",
      "\n",
      "#CPU threads: 16\n",
      "Scoring parameters: (Matrix=BLOSUM62 Lambda=0.267 K=0.041 Penalties=11/1)\n",
      "Temporary directory: \n",
      "#Target sequences to report alignments for: 25\n",
      "Opening the database...  [0s]\n",
      "Database: reference (type: Diamond database, sequences: 1, letters: 731)\n",
      "Block size = 2000000000\n",
      "Opening the input file...  [0.022s]\n",
      "Opening the output file...  [0s]\n",
      "Loading query sequences...  [0.439s]\n",
      "Masking queries...  [64.145s]\n",
      "Algorithm: Double-indexed\n",
      "Building query histograms...  [0.269s]\n",
      "Seeking in database...  [0s]\n",
      "Loading reference sequences...  [0s]\n",
      "Masking reference...  [0.001s]\n",
      "Initializing temporary storage...  [0.005s]\n",
      "Building reference histograms...  [0s]\n",
      "Allocating buffers...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 1/8, index chunk 1/4.\n",
      "Building reference seed array...  [0.002s]\n",
      "Building query seed array...  [0.076s]\n",
      "Computing hash join...  [0.01s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.04s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 1/8, index chunk 2/4.\n",
      "Building reference seed array...  [0.001s]\n",
      "Building query seed array...  [0.083s]\n",
      "Computing hash join...  [0.017s]\n",
      "Masking low complexity seeds...  [0s]\n",
      "Searching alignments...  [0.038s]\n",
      "Deallocating memory...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 1/8, index chunk 3/4.\n",
      "Building reference seed array...  [0.001s]\n",
      "Building query seed array... "
     ]
    }
   ],
   "source": [
    "# run minimap2\n",
    "\n",
    "result_dictionary = {}\n",
    "\n",
    "for i in sequence_dict:\n",
    "    this_ref = str(i) + \".ref.fasta\"\n",
    "\n",
    "    # note: this is the minimap2 command used in CZ ID short-read-mngs workflow\n",
    "    ! diamond makedb --in {this_ref} -d reference\n",
    "    \n",
    "    # note: there was an error when I was running with PE reads separately, \n",
    "    #       \"Unequal number of sequences in paired read files\" -- so I concatenate them above and run \n",
    "    #       the combined file through DIAMOND\n",
    "    ! diamond blastx -d reference -q miseq_reads_R1R2.fastq --mid-sensitive --outfmt 101 -o aln.sam\n",
    "    unmapped_reads_string = !grep -v ^@ aln.sam| cut -f3 | grep \"\\*\" | wc -l\n",
    "    unmapped_reads_count = int(unmapped_reads_string[0].strip())\n",
    "    \n",
    "    result_dictionary[i] = unmapped_reads_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ser = pd.Series(result_dictionary)\n",
    "print(ser)\n",
    "print(ser/100000)\n",
    "\n",
    "graph = (ser/100000).plot(kind='bar', title='% of reads unmapped -- ' + filename.split('/')[-1], \n",
    "                          ylabel = \"% of reads\", xlabel = \"%ID\",\n",
    "                         ylim = (0,1))\n",
    "graph.axhline(0.01, lw=.7, color='red')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
